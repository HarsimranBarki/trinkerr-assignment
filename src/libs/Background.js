const Background = ({ children }) => {
  return (
    // Remove transition-all to disable the background color transition.
    <div className="bg-white dark:bg-gray-800 dark:text-gray-100 h-screen transition-all">
      {children}
    </div>
  );
};
export default Background;
