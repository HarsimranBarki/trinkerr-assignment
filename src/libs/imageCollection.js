import one from "../images/one.png";
import two from "../images/two.png";
import three from "../images/three.png";
import four from "../images/four.png";
import five from "../images/five.png";

export const imageCollection = [
  {
    id: 1,
    name: "one",
    imageUrl: one,
    swiped: false,
  },
  {
    id: 2,
    name: "two",
    imageUrl: two,
    swiped: false,
  },
  {
    id: 3,
    name: "three",
    imageUrl: three,
    swiped: false,
  },
  {
    id: 4,
    name: "four",
    imageUrl: four,
    swiped: false,
  },
  {
    id: 5,
    name: "five",
    imageUrl: five,
    swiped: false,
  },
];
