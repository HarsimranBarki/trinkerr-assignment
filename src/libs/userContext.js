import React, { useMemo, useState } from "react";

export const UserContext = React.createContext(null);

export const UserProvider = ({ children }) => {
  const [user, setUser] = useState(
    JSON.parse(localStorage.getItem("user")) || null
  );
  const providerValue = useMemo(() => ({ user, setUser }), [user, setUser]);

  return (
    <UserContext.Provider value={providerValue}>
      {children}
    </UserContext.Provider>
  );
};
