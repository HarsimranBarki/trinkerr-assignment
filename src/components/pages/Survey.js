import { motion } from "framer-motion";
import React, { useContext, useEffect, useState } from "react";
import { useHistory } from "react-router-dom";
import { UserContext } from "../../libs/userContext";
import ImageCard from "../card/ImageCard";
import Header from "../common/Header";
import { imageCollection } from "../../libs/imageCollection";
import toast, { Toaster } from "react-hot-toast";

const Survey = () => {
  const { user, setUser } = useContext(UserContext);
  let history = useHistory();
  const [swipedCount, setSwipeCount] = useState(user.totalSwiped);
  const [currentImage, setCurrentImage] = useState(() => {
    if (swipedCount === 0) return imageCollection[0];
    if (swipedCount !== 5) return imageCollection[swipedCount];
    return null;
  });
  const toastGenerate = (message) => toast.success(message);

  const variants = {
    visible: {
      opacity: 1,
      x: 0,
      duration: 1,
      display: "block",
    },
    hidden: {
      opacity: 0,
      x: -10,
      transitionEnd: {
        display: "none",
      },
    },
  };

  const handleImage = (direction) => {
    if (swipedCount === 5) return;
    let message = "";
    if (direction === "left") message = "selected";
    if (direction === "right") message = "rejected";
    if (direction === "skipped") message = "skipped";

    toastGenerate(`${user.name} you have ${message} the image`);
    setSwipeCount((prev) => prev + 1);
    setUser({ ...user, totalSwipedSwiped: swipedCount + 1 });
    updateLocalStorage();
    setCurrentImage(imageCollection[swipedCount + 1]);
  };

  const updateLocalStorage = () => {
    let localUser = JSON.parse(localStorage.getItem("user"));
    let localCollection = JSON.parse(localStorage.getItem("userCollection"));

    localUser.totalSwiped = swipedCount + 1;
    localCollection.map((e) => {
      if (e.phone === user.phone) e.totalSwiped = swipedCount + 1;
    });

    localStorage.setItem("user", JSON.stringify(localUser));
    localStorage.setItem("userCollection", JSON.stringify(localCollection));
  };

  const handleKeyPress = (event) => {
    if (event.keyCode === 37) {
      event.stopPropagation();
      handleImage("left");
    } else if (event.keyCode === 39) {
      event.stopPropagation();
      handleImage("right");
    } else {
      return null;
    }
  };

  useEffect(() => {
    if (!user) history.push("/");
    let timer = setTimeout(() => {
      handleImage("skipped");
    }, 5000);
    window.addEventListener("keydown", handleKeyPress);
    return () => {
      window.removeEventListener("keydown", handleKeyPress);
      clearTimeout(timer);
    };
  }, [user, history]);

  return (
    <div className="p-10">
      <Header />
      <Toaster position="bottom-center" reverseOrder={false} />
      <div className="container max-w-screen-2xl m-auto flex  flex-wrap py-10">
        <motion.div
          className="flex-1"
          initial="hidden"
          animate="visible"
          transition={{ duration: 0.5, delay: 0.1 }}
          variants={variants}
        >
          <h2 className="text-4xl font-semibold">Hi {user?.name}</h2>
          <p className="text-sm mt-1">Welcome to the survey</p>
          {currentImage ? (
            <ImageCard url={currentImage.imageUrl} name={currentImage.name} />
          ) : (
            <p className="mt-10 text-xl">
              {user.name}, you have rated all the images. Thank You!
            </p>
          )}
        </motion.div>
        <motion.div
          className="flex-1  bg-blue-100 rounded p-5"
          initial="hidden"
          animate="visible"
          transition={{ duration: 0.5, delay: 0.2 }}
          variants={variants}
        >
          <h3 className="text-2xl font-medium">Instructions</h3>
          <div className="flex font-semobold flex-col space-y-1 mt-3 text-sm ">
            <p>
              Press Left Arrow Or Swipe Left To{" "}
              <span className="text-green-800">Accept</span>
            </p>
            <p>
              Press Rigth Arrow Or Swipe Right To{" "}
              <span className="text-red-800">Reject</span>
            </p>
            <p>Images Will Autoskip If Not Interacted</p>
          </div>
        </motion.div>
      </div>
    </div>
  );
};

export default Survey;
