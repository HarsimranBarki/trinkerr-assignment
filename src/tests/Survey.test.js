import { cleanup, render, screen } from "@testing-library/react";
import Adapter from "@wojtekmaj/enzyme-adapter-react-17";
import Enzyme, { mount, shallow } from "enzyme";
import toJson from "enzyme-to-json";
import React from "react";
import Survey from "../components/pages/Survey";
import { UserContext, UserProvider } from "../libs/userContext";

Enzyme.configure({ adapter: new Adapter() });

afterEach(cleanup);

describe("Survey", () => {
  it("Survey: should render correctly", () => {
    const wrapper = shallow(
      <UserProvider>
        <Survey />
      </UserProvider>
    );

    expect(toJson(wrapper)).toMatchSnapshot();
  });

  it("setSwipeCount: totalSwiped is set to 0 render Image", () => {
    const setUser = jest.fn();
    const user = { name: "Robin", phone: "1298787867", totalSwiped: 0 };
    render(
      <UserContext.Provider value={{ user, setUser }}>
        <Survey />
      </UserContext.Provider>
    );

    expect(screen.getByTestId("cardImage")).toMatchSnapshot();
  });

  it("setSwipeCount: totalSwiped is set to 5 don't render Image", () => {
    const setUser = jest.fn();
    const user = { name: "Robin", phone: "1298787867", totalSwiped: 5 };
    render(
      <UserContext.Provider value={{ user, setUser }}>
        <Survey />
      </UserContext.Provider>
    );

    expect(screen.getByTestId("thank-you-message")).toMatchSnapshot();
  });

  it("refresh: state is preserved with correct Image", () => {
    const setUser = jest.fn();
    const user = { name: "Robin", phone: "1298787867", totalSwiped: 3 };
    localStorage.setItem("user", JSON.stringify(user));

    render(
      <UserContext.Provider value={{ user, setUser }}>
        <Survey />
      </UserContext.Provider>
    );

    expect(screen.getByTestId("cardImage").id).toBe("4");
  });
});
