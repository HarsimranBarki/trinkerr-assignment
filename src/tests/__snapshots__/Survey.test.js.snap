// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Survey Survey: should render correctly 1`] = `
<ContextProvider
  value={
    Object {
      "setUser": [Function],
      "user": null,
    }
  }
>
  <Survey />
</ContextProvider>
`;

exports[`Survey setSwipeCount: totalSwiped is set to 0 render Image 1`] = `
<img
  alt="one"
  class="  max-h-sm w-100 sm:max-w-sm  bg-blue-100 p-5 mt-5 shadow-md swipe-image"
  data-testid="cardImage"
  draggable="false"
  id="1"
  name="one"
  src="one.png"
  style="opacity: 1; transform: none; user-select: none;"
/>
`;

exports[`Survey setSwipeCount: totalSwiped is set to 5 don't render Image 1`] = `
<p
  class="mt-10 text-xl"
  data-testid="thank-you-message"
  id="thank-you-message"
>
  Robin
  , you have rated all the images. Thank You!
</p>
`;
